### PR TITLE
perf: replace Substring with AsSpan slicing in hot paths

### LINKED ISSUE #501: perf: Replace Substring() with AsSpan() slicing on per-record hot paths
## Summary

Several production files still use `.Substring()` to slice strings, which allocates a new `string` on every call. On per-record hot paths this means one (or more) unnecessary heap allocations per generated document. These can be replaced with `.AsSpan()` slicing or `string.Concat(ReadOnlySpan<char>, ...)` overloads for zero-allocation alternatives.

Zipper already uses `.AsSpan()` in `ChaosEngine.cs` (9 existing call sites) and `stackalloc` in `TextGenerator.cs` — this is about closing the remaining gaps.

### Hot path (per-record / per-document) — worth fixing

| File | Line | Code | Frequency |
|------|------|------|-----------|
| `src/ChaosEngine.cs` | 496 | `line.Substring(firstComma)` | Per affected record |
| `src/LoadFiles/OptComposer.cs` | 129 | `baseImagePath.Substring(0, baseImagePath.Length - ext.Length)` | Per multi-page document |
| `src/ProductionSetGenerator.cs` | 119 | `plan.ImageRelPath.Substring(0, plan.ImageRelPath.Length - imageExt.Length)` | Per multi-page document |
| `src/Utils/NamingConventionHelper.cs` | 63 | `word.Substring(1).ToLowerInvariant()` | Per word per column name |

### Cold path (CLI parsing, runs once) — not worth the churn

| File | Line | Code |
|------|------|------|
| `src/Cli/CliValidator.cs` | 493 | `value.Substring(6)` |
| `src/Cli/RequestBuilder.cs` | 190, 269, 280 | Various arg parsing substrings |

## Suggested Approach

For each hot-path site, replace `str.Substring(start, len)` with the appropriate zero-allocation alternative:

- **Strip extension pattern** (`OptComposer`, `ProductionSetGenerator`): Use `Path.GetFileNameWithoutExtension()` or `path.AsSpan()[..^ext.Length]` depending on whether the result needs to be a `string` for interpolation.
- **Slice-from-index** (`ChaosEngine`): `string.Concat(line.AsSpan(firstComma))` or use `AsSpan` directly if the result feeds into another `string.Concat` (which ChaosEngine already does on adjacent lines).
- **ToLowerInvariant on tail** (`NamingConventionHelper`): `word.AsSpan(1)` into a `Span<char>` buffer with `ToLowerInvariant`, or check if the `StringBuilder` overload accepts `ReadOnlySpan<char>` directly (it does in .NET 10).

Leave the CLI parsing cold-path calls alone — not worth the churn for code that runs once per invocation.

## Verification

```bash
dotnet build zipper.sln && dotnet format --verify-no-changes src/ && dotnet test src/Zipper.Tests/Zipper.Tests.csproj
```

Golden parity check recommended since `ChaosEngine` and `OptComposer` affect output.

### PR DESCRIPTION
**Pull request by dwojtaszek**
Closes #501

## Release Notes
Replaced string Substring allocations with zero-allocation AsSpan slicing in hot paths and optimized NamingConventionHelper's ToPascalCase with a stack-allocated buffer to reduce overall memory allocations.

## Summary by CodeRabbit

* **Refactor**
  * Optimized string slicing operations in file processing components for improved performance
  * Enhanced memory efficiency in naming convention transformations through optimized buffer allocation
  
* **Tests**
  * Added test coverage for edge-case scenarios in file processing and naming operations
  * Added performance validation tests for naming convention operations

### CODE REVIEW

File: src/ChaosEngine.cs [hunk 1]
```diff
@@ -493,7 +493,7 @@ private static string ApplyOptBatesNumberCorruption(string line)
         int firstComma = line.IndexOf(',', StringComparison.Ordinal);
         if (firstComma >= 0)
         {
-            return line.Substring(firstComma); // Leaves empty Bates Number
+            return new string(line.AsSpan(firstComma)); // Leaves empty Bates Number
```
[the remark this answers is not in the document]
Comment (dwojtaszek) [reply src/ChaosEngine.cs]: Fixed. Reverted to Substring(firstComma) to avoid redundant new string(AsSpan) allocation.

File: src/LoadFiles/OptComposer.cs [hunk 2]
```diff
@@ -125,14 +125,11 @@ private IEnumerable<LoadFileRecord> ComposeFromFiles(IReadOnlyList<FileData> pro
...
             var ext = Path.GetExtension(baseImagePath);
-            var pathWithoutExt = baseImagePath.Length >= ext.Length
-                ? baseImagePath.Substring(0, baseImagePath.Length - ext.Length)
-                : baseImagePath;
 
             for (int pageIdx = 1; pageIdx <= actualPages; pageIdx++)
             {
                 var pageBates = $"{baseBates}_{pageIdx:D3}";
-                var pageImagePath = $"{pathWithoutExt}_{pageIdx:D3}{ext}";
+                var pageImagePath = $"{baseImagePath.AsSpan()[..^ext.Length]}_{pageIdx:D3}{ext}";
```
[the remark this answers is not in the document]
Comment (dwojtaszek) [reply src/LoadFiles/OptComposer.cs:132]: Skip reason: Slicing the span inside the loop is necessary to avoid compilation error CS4007 ('Instance of type System.ReadOnlySpan<char> cannot be preserved across await or yield boundary'). Because this is an iterator/async method, a local variable System.ReadOnlySpan<char> declared outside the loop cannot be preserved in the generated state machine. Slicing the span inline within the string interpolation statement is a ref struct stack operation and has zero heap allocation overhead, which correctly optimizes the hot path without compilation failure.

File: src/ProductionSetGenerator.cs [hunk 3]
```diff
@@ -115,13 +115,10 @@ private static async Task<ProductionSetResult> GenerateCoreAsync(
...
             {
                 var imageExt = Path.GetExtension(plan.ImageRelPath);
-                var imagePathWithoutExt = plan.ImageRelPath.Length >= imageExt.Length
-                    ? plan.ImageRelPath.Substring(0, plan.ImageRelPath.Length - imageExt.Length)
-                    : plan.ImageRelPath;
 
                 for (int pageIdx = 1; pageIdx <= generated.PageCount; pageIdx++)
                 {
-                    var pageImageRelPath = $"{imagePathWithoutExt}_{pageIdx:D3}{imageExt}";
+                    var pageImageRelPath = $"{plan.ImageRelPath.AsSpan()[..^imageExt.Length]}_{pageIdx:D3}{imageExt}";
```
[the remark this answers is not in the document]
Comment (dwojtaszek) [reply src/ProductionSetGenerator.cs:121]: Skip reason: Slicing the span inside the loop is necessary to avoid compilation error CS4007 ('Instance of type System.ReadOnlySpan<char> cannot be preserved across await or yield boundary'). Because this is an iterator/async method, a local variable System.ReadOnlySpan<char> declared outside the loop cannot be preserved in the generated state machine. Slicing the span inline within the string interpolation statement is a ref struct stack operation and has zero heap allocation overhead, which correctly optimizes the hot path without compilation failure.
